Public/Src/Tools/SandboxedProcessExecutor/OutputErrorObserver.cs

```diff
@@ -0,0 +1,126 @@
+            // To avoid that, only scan the first 400 characters. That's enough for
+            // the longest possible prefix: MAX_PATH, plus a huge subcategory string, and an error location.
+            // After the regex is done, we can append the overflow.
+            if (line.Length > 400)
+            {
+                line = line.Substring(0, 400);
+            }
+
+            // If a tool has a large amount of output that isn't a warning (eg., "dir /s %hugetree%")
+            // the regex matching below may be slow. It's faster to pre-scan for "warning"
+            // and bail out if neither are present.
+            if (m_isDefaultWarningRegex &&
+                line.IndexOf("warning", StringComparison.OrdinalIgnoreCase) == -1)
+            {
+                return false;
+            }
+
+            return m_warningRegex.IsMatch(line);
+        }
+
+        /// <summary>
+        /// Creates an instance of <see cref="OutputErrorObserver"/>
+        /// </summary>
+        public static OutputErrorObserver Create(SandboxedProcessInfo info)
+        {
+            Contract.Requires(info != null);
+
+            Regex warningRegex = info.StandardObserverDescriptor == null || info.StandardObserverDescriptor.WarningRegex == null
+                ? null
+                : CreateRegex(info.StandardObserverDescriptor.WarningRegex);
+            bool isDefaultWarningRegex = warningRegex == null ? false : IsDefaultWarningRegex(info.StandardObserverDescriptor.WarningRegex);
+            bool logOutputToConsole = info.StandardObserverDescriptor == null ? false : info.StandardObserverDescriptor.LogOutputToConsole;
+            bool logErrorToConsole = info.StandardObserverDescriptor == null ? false : info.StandardObserverDescriptor.LogErrorToConsole;
+
+            return new OutputErrorObserver(warningRegex, isDefaultWarningRegex, logOutputToConsole, logErrorToConsole);
+        }
+
+        private static bool IsDefaultWarningRegex(SandboxedProcessInfo.RegexDescriptor regexDescriptor) 
+            => string.Equals(regexDescriptor.Pattern, Warning.DefaultWarningPattern) && regexDescriptor.Options == RegexOptions.IgnoreCase;
+
+        private static Regex CreateRegex(SandboxedProcessInfo.RegexDescriptor regexDescriptor) 
+            => new Regex(
+                    regexDescriptor.Pattern,
+                    regexDescriptor.Options | RegexOptions.Compiled | RegexOptions.CultureInvariant);
+
+        /// <summary>
+        /// Observes output for warning.
+        /// </summary>
+        public void ObserveOutput(string outputLine)
+        {
+            if (IsWarning(outputLine))
+            {
+                Interlocked.Increment(ref m_warningCount);
+            }
+
+            if (m_logOutputToConsole)
+            {
+                Console.Out.WriteLine(outputLine);
+            }
+        }
+
+        /// <summary>
+        /// Observes error for warning.

```

Comment:
>warning [](start = 31, length = 7)

error

Reply:
No we observed standard error for warning.

---
In reply to: [276485208](https://github.com/Microsoft/BuildXL/pull/159#discussion_r276485208) [](ancestors = 276485208)

Reply:
you're right

---
In reply to: [276870633](https://github.com/Microsoft/BuildXL/pull/159#discussion_r276870633) [](ancestors = 276870633,276485208)